Downsample/Downsample.cpp

```diff
@@ -138,17 +138,26 @@ Eigen::MatrixXf gridSum(const Eigen::MatrixXf &r, int downsamplingFactor)
 
 Eigen::MatrixXf lineSum(const Eigen::MatrixXf &matrix, int downsamplingFactor)
 {
+    Eigen::Map<Eigen::MatrixXf> reshapedMatrixMap(const_cast<float *>(matrix.data()),
+                                                  downsamplingFactor,
+                                                  (int)matrix.rows() * (int)matrix.cols() / downsamplingFactor);
     Eigen::MatrixXf reshapedMatrix =
-        matrix.reshaped(downsamplingFactor, (int)matrix.rows() * (int)matrix.cols() / downsamplingFactor);
+        Eigen::Map<Eigen::MatrixXf>(reshapedMatrixMap.data(), reshapedMatrixMap.rows(), reshapedMatrixMap.cols());
 
     std::function<float(float)> nan_to_zero_functor = nanToZero;
 
     reshapedMatrix = reshapedMatrix.unaryExpr(nan_to_zero_functor);
 
-    return reshapedMatrix.colwise()
-        .sum()
-        .reshaped((int)matrix.rows() / downsamplingFactor, (int)matrix.cols())
-        .transpose();
+    Eigen::MatrixXf reshapedMatrixColwiseSum = reshapedMatrix.colwise().sum();
+
+    Eigen::Map<Eigen::MatrixXf> reshapedMatrixMap2(reshapedMatrixColwiseSum.data(),
+                                                   (int)matrix.rows() / downsamplingFactor,
+                                                   (int)matrix.cols());
+
+    Eigen::MatrixXf reshapedMatrix2 =
+        Eigen::Map<Eigen::MatrixXf>(reshapedMatrixMap2.data(), reshapedMatrixMap2.rows(), reshapedMatrixMap2.cols());
+
+    return reshapedMatrix2.transpose();
```

Comment:
I don't know Eigen, how does `transpose` work? The docs say it returns "an expression of the transpose of *this.", whatever that means. Sounds like it's some sort of lazy thing which refers back to `reshapedMatrix2` which goes out of scope when the function exits? Is that safe?

Reply:
Eigen::MatrixXf reshapedMatrix =
        Eigen::Map<Eigen::MatrixXf>(reshapedMatrixMap.data(), reshapedMatrixMap.rows(), reshapedMatrixMap.cols())
            .transpose();

    return reshapedMatrix;

Is this better?

Reply:
I don't know. Someone who knows Eigen has to figure this out.

Reply:
.transpose has been moved out of the lineSum()